plugins/modules/source_control/gitlab/gitlab_service.py

```diff
@@ -0,0 +1,468 @@
+    default: ["push", "issues", "confidential_issues", "merge_requests", "tag_push", "note", "confidential_note", "job", "pipeline", "wiki_page"]
+    type: list
+    elements: str
+  state:
+    description:
+      - Create or delete a service.
+      - Possible values are C(present) and C(absent).
+    default: "present"
+    choices: ["present", "absent"]
+    type: str
+'''
+
+EXAMPLES = '''
+# Setup email on push for this project
+- name: emails me on push
+  community.general.gitlab_service:
+    api_url: https://gitlab.com
+    api_token: foobar
+    project: 123456
+    service: emails-on-push
+    params:
+      recipients: foo@example.com
+      disable_diffs: true
+    state: present
+  delegate_to: localhost
+
+# This will always be set to change because a non-null token is mandatory
+- name: trigger packagist update on push events (only)
+  community.general.gitlab_service:
+    api_url: https://gitlab.com
+    api_token: foobar
+    project: foo/proj
+    service: packagist
+    events: ["push"]
+    params:
+      username: foo
+      token: bar
+      server: https://packagist.org
+  delegate_to: localhost
+
+- Idempotency is only partially provided since GitLab does
+  not expose secret parameters like tokens or passwords.
+  See U(https://gitlab.com/gitlab-org/gitlab-ce/issues/46313)
+'''
+
+RETURN = '''
+---
+service:
+  description: A dict containing key/value pairs representing GitLab service
+  returned: success
+  type: dict
+  sample:
+    id: 40812345
+    push_events: true
+    issues_events: true
+    confidential_issues_events: true
+    merge_requests_events: true
+    tag_push_events: true
+    note_events: true
+    confidential_note_events: true
+    job_events: true
+    pipeline_events: true
+    wiki_page_events: true
+    recipients: me@example.com
+    disable_diffs: null
+    send_from_committer_email: null
+    title: Emails on push
+    created_at: '2018-05-13T03:08:07.943Z'
+    updated_at: '2018-05-13T03:09:17.651Z'
+    active: true
+    properties:
+      send_from_committer_email: null
+      disable_diffs: null
+      recipients: me@example.com
+    project_id: 1234567
+state:
+  description: A string indicating whether the service was "created" or "changed"
+  returned: success
+  type: str
+  sample: created
+'''
+
+import traceback
+
+from ansible.module_utils.six import iteritems
+
+GITLAB_IMP_ERR = None
+try:
+    import gitlab
+    HAS_GITLAB_PACKAGE = True
+except ImportError:
+    GITLAB_IMP_ERR = traceback.format_exc()
+    HAS_GITLAB_PACKAGE = False
+
+from ansible.module_utils.api import basic_auth_argument_spec
+from ansible.module_utils.basic import AnsibleModule, missing_required_lib
+from ansible.module_utils._text import to_native
+
+from ansible_collections.community.general.plugins.module_utils.gitlab import findProject, gitlabAuthentication
+
+# auto-generated 2020/07/09 from https://gitlab.com/gitlab-org/gitlab/-/blob/master/lib/api/helpers/services_helpers.rb
```

Comment:
How is it generated from that source, i.e. how can this definition be updated when the source changes?

Reply:
Download said file,
prepend:
```ruby
require 'json'
class Boolean
end
```

append the following code block:
```ruby
no_log_keys = [:api_key, :webhook, :token, :password]
services.each do |service_slug, params|
  new_settings = {}

  params.each do |entry|
    e = {}

    if (entry[:required] === true)
      e[:required] = 1
    end
    if (entry[:type] == String)
      e[:type] = "str"
    elsif (entry[:type] == Integer)
      e[:type] = "int"
    elsif (entry[:type] == Boolean)
      e[:type] = "bool"
    else
      e[:type] = entry[:type]
    end

    if (no_log_keys.find_index(entry[:name])) != nil
      e[:no_log] = true
    end
    # e[:desc] = entry[:desc]

    new_settings[entry[:name]] = e
  end

  services[service_slug] = new_settings
end

services = Hash[*services.sort.flatten]

puts JSON.generate(services, options = {
  indent:'',
  space: ' '
}).gsub(/}},/, "}},\n").gsub(/,(?![ ])/, ', ').gsub(/\btrue\b/, 'True')
```

And run using the `ruby` interpreter. (Some better way probably exist)
(I also pipe to my little `wrap-on-sequence -w 130 "}," ","` tool in order to wrap/fold it nicely)

Reply:
Can you add that as a comment in the code, so that people will be able to find this information when they want to update it?